sound/soc/sof/ipc4-topology.c

```diff
@@ -1720,6 +1721,16 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 				alh_copier = (struct sof_ipc4_copier *)dai->private;
 				alh_data = &alh_copier->data;
 				blob->alh_cfg.mapping[i].device = alh_data->gtw_cfg.node_id;
+
+				/*
+				 * The dma_config_tlv[i] mapping device should be the same as
+				 * the mapping[i] device in ALH blob. The device id is used
+				 * only for DMA tlv mapping purposes.
+				 */
+				dma_config = &ipc4_copier->dma_config_tlv[i].dma_config;
+				dma_config->dma_stream_channel_map.mapping[0].device =
+					blob->alh_cfg.mapping[i].device;
```

Comment:
I keep repeating the same question "what is a device" in this context?

It's too cryptic and can be explained better.

Reply:
@plbossart Currently, the `device` = node_id which is `node_type | stream_tag - 1`. But, it doesn't have to be `stream_tag - 1`. The only rule is that the `device` is unique for each stream and should be identical between dma_config and ALH blob. @iganakov  Can you advice what `device` should be?

Reply:
The comments in the firmware code seem to suggest it was link_id | pdi_id?

I'd really like to have a clear documented description of what this argument needs to be, otherwise this could be fun with breakages at any time.

Reply:
> The comments in the firmware code seem to suggest it was link_id | pdi_id?

Sorry, I didn't see that comment. And I notice that the dai_index is stream_tag. https://github.com/thesofproject/sof/pull/8361#discussion_r1467696549
Given that there is no ALH hardware on ACE2.x, I think it is more reasonable to use stream_tag?

BTW, I can add a commit to use `link_id | pdi_id` if we are all agree.

> 
> I'd really like to have a clear documented description of what this argument needs to be, otherwise this could be fun with breakages at any time.

Totally agree.

Reply:
Can you check with @iganakov and document all this?

Reply:
there's another comment which mentions alh_id

https://github.com/thesofproject/sof/pull/8361/files#diff-11acf7917806ca7b7727d4b030f8b7ff71c153d4e5b0b757d32e59ed2ea0ef37R1065

````c
		/* To be able to retrieve proper DMA config we need to check if
		 * device_id value (which is alh_id) is equal to device_address.
		 * They both contain SNDW master id and PDI. If they match then
		 * proper config is found.
````

but it's unclear what the encoding might be... 

Now that the fw part has been merged, we have a regression filed, so we need to progress on this PR and bring back the functionality.

Reply:
> there's another comment which mentions alh_id
> 
> https://github.com/thesofproject/sof/pull/8361/files#diff-11acf7917806ca7b7727d4b030f8b7ff71c153d4e5b0b757d32e59ed2ea0ef37R1065
> 
> ```c
> 		/* To be able to retrieve proper DMA config we need to check if
> 		 * device_id value (which is alh_id) is equal to device_address.
> 		 * They both contain SNDW master id and PDI. If they match then
> 		 * proper config is found.
> ```
> 
> but it's unclear what the encoding might be...
> 
> Now that the fw part has been merged, we have a regression filed, so we need to progress on this PR and bring back the functionality.

@plbossart I think `alh_id` here is referring to `alh_blob->alh_cfg.mapping[i].alh_id`. That is what I mentioned the device should be identical between dma_config and ALH blob. So, it is `dma_cfg->channel_map.map[i].device_address = alh_blob->alh_cfg.mapping[i].alh_id` in the firmware. And the device_address (alh_id) can be any number. @iganakov Please correct me if I am wrong.

Reply:
@bardliao Yes, you're right, `dma_cfg->channel_map.map[i].device_address = alh_blob->alh_cfg.mapping[i].alh_id`. And from the parsing algorithm perspective device_address (alh_id) can be any number

Reply:
@iganakov @bardliao can we try to align with the other OS that shall not be named so that we use the same convention for the device address. This is one of the cases where we don't want to be creative, using the same convention is useful when comparing IPC payloads.